### PR TITLE
Testcase big document #180

### PR DESCRIPTION
**Pull request by rototor**
Related to #180 

This takes 5m 49s on my MacBook Pro 2014 to create a HTML file with 18.5 MB and a result PDF with 232 MB and 12694 pages using JDK 1.7.0_52.

Feel free to change the list directive
```freemarker
[#list 1..5000 as page] 
```
from 5000 to something smaller.

### CODE REVIEW

File: openhtmltopdf-core/src/main/java/com/openhtmltopdf/context/StyleReference.java [hunk 1]
```diff
@@ -251,22 +249,22 @@ private List getStylesheets() {
...
                     Stylesheet sheet = _stylesheetFactory.parse(
-                            new StringReader(refs[i].getContent()), refs[i]);
-                    refs[i].setStylesheet(sheet);
-                    refs[i].setUri(null);
+                            new StringReader(ref.getContent()), ref);
+                    ref.setStylesheet(sheet);
+                    ref.setUri(null);
                 }
             }
+            infos.addAll(Arrays.asList(refs));
```
Comment (rototor) [inline openhtmltopdf-core/src/main/java/com/openhtmltopdf/context/StyleReference.java:266]: This is something IDEA cached ... if refs == null, the addAll would throw an NPE...

File: openhtmltopdf-core/src/main/java/com/openhtmltopdf/css/newmatch/CascadedStyle.java [hunk 2]
```diff
@@ -227,23 +227,17 @@ public IdentValue getIdent(CSSName cssName) {
...
      */
-    public java.util.Iterator getCascadedPropertyDeclarations() {
-        List list = new ArrayList(cascadedProperties.size());
-        Iterator iter = cascadedProperties.values().iterator();
-        while ( iter.hasNext()) {
-            list.add(iter.next());
-        }
-        return list.iterator();
+    public java.util.Collection<PropertyDeclaration> getCascadedPropertyDeclarations() {
+        return cascadedProperties.values();
```
Comment (rototor) [inline openhtmltopdf-core/src/main/java/com/openhtmltopdf/css/newmatch/CascadedStyle.java:231]: This collection here is always directly consumed and iterated over. There was no sense in copying this...

File: openhtmltopdf-examples/src/main/resources/freemarker/many_pages.ftl [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+	<li><a href="#sa${sectionAnchorCounter?c}">${content}</a></li>
+	[/#assign]
+[/#macro]
+[#macro h1][@hn 1][#nested][/@hn][/#macro]
+[#macro h2][@hn 2][#nested][/@hn][/#macro]
+[#macro h3][@hn 3][#nested][/@hn][/#macro]
+[#macro h4][@hn 4][#nested][/@hn][/#macro]
+
+
+[#list 1..100 as page]
```
Comment (rototor) [inline openhtmltopdf-examples/src/main/resources/freemarker/many_pages.ftl]: increase this count here for way more pages

File: openhtmltopdf-examples/src/main/resources/freemarker/many_pages.ftl [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+				AFECTADO POR
+			</th>
+			<th class="hdrAsig" style="border-bottom: 1px solid">
+				PAGADO
+			</th>
+		</tr>
+		</thead>
+
+		<tbody>
+			[#list 1..1000 as subRow]
```
Comment (rototor) [inline openhtmltopdf-examples/src/main/resources/freemarker/many_pages.ftl]: This count is high at the moment, before merging this branch it should be decreased to a sensible value, as travis won't finish with such many rows...

**Comment by danfickle** [comment]
I’m going to work on style caching now. Provided there are no sibling selectors or custom nth-child I think we can do a much better job at avoiding running the cascade for every element.

**Pull request merged**